### PR TITLE
offchain - libraries improvements

### PR DESCRIPTION
**Pull request by dimkouv**
Library code improvements

- Use existing `mathutil.Max` and `utils.ZeroAddress`
- Create `contractutil` `logpollerutil` 
- Place `plugins_common.go` functions under `internal/` libs.
- Delete `rpc_helpers.go` and Move `LazyFetch` under `cache/`

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/internal/logpollerutil/filters.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+func RegisterLpFilters(lp logpoller.LogPoller, filters []logpoller.Filter, qopts ...pg.QOpt) error {
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/ccip/internal/logpollerutil/filters.go:11]: Food for thought. I don't know what path we want to follow for functionality extensions to the interfaces we import. Although very imperative, this approach is fine. On the other hand, I was thinking about a more object-oriented approach in which we can encapsulate those functions within a single type. This is how I used to do it in Java, but Go for cases like this is more flexible. As I mentioned, not saying the approach is better. don't know to be honest. Just food for thought ;) 

```go
type LogPollerAdapter interface {
	logpoller.LogPoller
	RegisterLpFilters(filters []logpoller.Filter, qopts ...pg.QOpt) error
	UnregisterLpFilters(filters []logpoller.Filter, qopts ...pg.QOpt) error
}

type logPollerAdapter struct {
	logpoller.LogPoller 
}

func NewLogPollerAdapter(lp logpoller.LogPoller) LogPollerAdapter {
	return &logPollerAdapter{lp}
}

func (lp *logPollerAdapter) RegisterLpFilters(filters []logpoller.Filter, qopts ...pg.QOpt) error {
	for _, lpFilter := range filters {
		if filterContainsZeroAddress(lpFilter.Addresses) {
			continue
		}
		if err := lp.RegisterFilter(lpFilter, qopts...); err != nil {
			return err
		}
	}
	return nil
}

func (lp *logPollerAdapter) UnregisterLpFilters(filters []logpoller.Filter, qopts ...pg.QOpt) error {
	for _, lpFilter := range filters {
		if filterContainsZeroAddress(lpFilter.Addresses) {
			continue
		}
		if err := lp.UnregisterFilter(lpFilter.Name, qopts...); err != nil {
			return err
		}
	}
	return nil
}
```

File: core/services/ocr2/plugins/ccip/internal/hashlib/intervals.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/commit_store"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/evm_2_evm_onramp"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/ccipcalc"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/ccipdata"
+)
+
+// LeavesFromIntervals Extracts the hashed leaves from a given set of logs
+func LeavesFromIntervals(
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/ccip/internal/hashlib/intervals.go]: I wonder if we don't have too much granularity here at this moment. Although I like more, but smaller files, I have a feeling that we don't have any good arguments to split it so much.

File: core/services/ocr2/plugins/ccip/internal/hashlib/common.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package hashlib
+
+import "github.com/smartcontractkit/chainlink/v2/core/utils"
+
+func BytesOfBytesKeccak(b [][]byte) ([32]byte, error) {
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/ccip/internal/hashlib/common.go:15]: Maybe all hashlb files merged into one?

**Review by mateusz-sekara — COMMENTED** [review]
Do we need such granularity at this moment? I'd instead invest more into splitting plugins into smaller, understandable chunks than some of the utils in this PR. I feel that maybe contractutil can be a single file, same for hashlib?

File: core/services/ocr2/plugins/ccip/internal/logpollerutil/filters.go (hunk 1 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/internal/logpollerutil/filters.go:11]: Yes I like this approach, but the purpose of this PR is to split the logic into packages/files and also test those.
Refactoring each one of those packages is going to be a follow-up.

File: core/services/ocr2/plugins/ccip/internal/hashlib/intervals.go (hunk 2 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/internal/hashlib/intervals.go]: IMO it's minor as long as everything is in the same package, putting everything into a single file is trivial.
I like keeping each file < 200 lines, loaders/shortcuts should be ok atm.

File: core/services/ocr2/plugins/ccip/internal/hashlib/common.go (hunk 3 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/internal/hashlib/common.go:15]: moved intervals into common

**Review by mateusz-sekara — APPROVED** [review]

**Review by RensR — APPROVED** [review]
bit more fine grained than I would have done myself but OK to me

**Pull request merged**